Mathlib/Algebra/Exact.lean

```diff
@@ -35,30 +36,106 @@ variable (f : M → N) (g : N → P)
   `g y = 0` iff `y` belongs to the image of `f` -/
 def Exact [Zero P] : Prop := ∀ y, g y = 0 ↔ y ∈ Set.range f
 
+variable {f g}
+
+lemma Exact.apply_apply_eq_zero [Zero P] (h : Exact f g) (x : M) :
+    g (f x) = 0 := (h _).mpr <| Set.mem_range_self _
+
 lemma Exact.comp_eq_zero [Zero P] (h : Exact f g) : g.comp f = 0 :=
-  funext fun _ => (h _).mpr <| Set.mem_range_self _
+  funext h.apply_apply_eq_zero
+
+lemma Exact.of_comp_eq_zero_of_eq_zero_imp_mem_range [Zero P]
+    (h1 : g ∘ f = 0) (h2 : ∀ x, g x = 0 → x ∈ Set.range f) : Exact f g :=
+  fun y => Iff.intro (h2 y) <|
+    Exists.rec ((forall_apply_eq_imp_iff (p := (g · = 0))).mpr (congrFun h1) y)
 
 end Function
 
 section LinearMap
 
 open LinearMap
 
-variable {R : Type*} [CommSemiring R]
-  [AddCommMonoid M] [AddCommMonoid N] [AddCommMonoid P]
-  [Module R M] [Module R N] [Module R P]
+variable [Semiring R] [AddCommMonoid M] [AddCommMonoid M'] [AddCommMonoid N]
+  [AddCommMonoid N'] [AddCommMonoid P] [AddCommMonoid P'] [Module R M]
+  [Module R M'] [Module R N] [Module R N'] [Module R P] [Module R P']
 
 variable {f : M →ₗ[R] N} {g : N →ₗ[R] P}
 
 lemma Exact.linearMap_ker_eq (hfg : Exact f g) : ker g = range f :=
   SetLike.ext hfg
 
-lemma LinearMap.exact_iff : Exact f g ↔ LinearMap.ker g = LinearMap.range f :=
+lemma _root_.LinearMap.exact_iff :
+    Exact f g ↔ LinearMap.ker g = LinearMap.range f :=
   Iff.symm <| SetLike.ext_iff
 
+lemma _root_.LinearMap.exact_of_comp_eq_zero_of_ker_le_range
+    (h1 : g ∘ₗ f = 0) (h2 : ker g ≤ range f) : Exact f g :=
+  Exact.of_comp_eq_zero_of_eq_zero_imp_mem_range (congrArg DFunLike.coe h1) h2
+
+lemma _root_.LinearMap.exact_of_comp_eq_zero_of_eq_zero_imp_mem_range
+    (h1 : g ∘ₗ f = 0) (h2 : ∀ x, g x = 0 → x ∈ range f) : Exact f g :=
+  exact_of_comp_eq_zero_of_ker_le_range h1 h2
+
 lemma Exact.linearMap_comp_eq_zero (h : Exact f g) : g.comp f = 0 :=
   DFunLike.coe_injective h.comp_eq_zero
 
+lemma Surjective.precomp_exact_iff_exact {p : M' →ₗ[R] M} (h : Surjective p) :
```

Comment:
I think we use `comp_right` instead of `precomp`? Or maybe even `Surjective.comp_exact_iff`.